## Pandas
- Pandas is a open source python library designed specifically for accomplishing data analysis tasks.
- When we look into source code of pandas, we  can verify that Pandas library is built on top of Numpy library.
- Pandas library provides two important data structures named:
    - [pandas.Series:](https://github.com/pandas-dev/pandas/blob/v2.0.3/pandas/core/series.py#L244-L6112)
        - Series is a one-dimensional ndarray with axis labels.
        - You can think of Series as a Single Column or Single Row of data with associated labels.

    - [`pandas.DataFrame:`](https://github.com/pandas-dev/pandas/blob/v2.0.3/pandas/core/frame.py#L490-L11586)
        - A DataFrame is a two-dimensional labeled data structure with columns that can hold data of different types.
        - It is similar to a table in a relational database or a spreadsheet with rows and columns.
        - Each column in a DataFrame can be visualize as Series.
        - A DataFrame can be created from dictionaries, arrays, list of dictionaries, CSV Files, Excel Files, and more.
        - Pandas DataFrame class consists of methods and attributes that makes the life of data analyst easier.

In addition to these Data Structures, Pandas also provides various tools for Data ingestion, Data manipulation, Data Merging, Grouping and so on.


## Install Pandas and Numpy
- Before using pandas, you need to make sure pandas libary is installed into your system.
- If you are using Google Colab, there is no need to install pandas as it comes with by default.
- In addition to Pandas, we will also use Numpy library, lets install this as well (skip if you are using Google Colab)
- Uncomment below cell to install necessary libraries

In [1]:
# !pip install pandas
# !pip install numpy

## Import Libraries
- In order to use advantange of installed libraries i.e. numpy and pandas, We need to import it using `import` keyword.
- We will represent:
    - numpy via alias `np`
    - pandas via alias `pd`

In [2]:
import numpy as np
import pandas as pd

## Creating Series
- There are several ways in creating a Series:
    1. From List
    2. From Dictionary
    3. From Numpy Array

- To create a a Series you can use `pandas.Series()` function.

**1. Create a Series from python list**

In [3]:
# define list
data = [10, 20, 30, 40, 50]

# create series from list
series_list = pd.Series(data)

print(series_list)

0    10
1    20
2    30
3    40
4    50
dtype: int64


In [4]:
# check type
print(type(series_list))

<class 'pandas.core.series.Series'>


**2. Create Series from Dictionary**

In [5]:
# define dictionary
data_dict = {
    'a': 10,
    'b': 20,
    'c': 30,
    'd': 40
}

# create series from dictionary
series_dict = pd.Series(data_dict)

print(series_dict)

a    10
b    20
c    30
d    40
dtype: int64


In [6]:
# check type
print(type(series_dict))

<class 'pandas.core.series.Series'>


**Create Dictionary from Numpy Array**

In [7]:
# define array
numpy_array = np.array([10, 20, 30, 40])

# create series from numpy array
series_np = pd.Series(numpy_array)

print(series_np)

0    10
1    20
2    30
3    40
dtype: int64


- **Note:**
    - When creating `series` via list and  numpy array, Series index label are automatically generated starting from 0.
    - When creating `Series` via dictionary, Dictionary keys becomes Series index label and Dictionary values becomes the data in Series.

## Create DataFrame
- Several ways to create Pandas DataFrame are:
    1. Create DataFrame from 2D List
    2. Create DataFrame a Dictionary of Lists
    3. Create DataFrame from a List of Dictionaries
    4. Create DataFrame from a CSV file
    5. Create DataFrame from an Excel file

- To create a DataFrame you can use `pd.DataFrame()` function.

**1. Create DataFrame from 2D List**

In [8]:
# Define list
data = [[10, 20], [30, 40], [50, 60]]
# print(len(data)) -> 60

# create DataFrame
df_list = pd.DataFrame(data, columns=['col1', 'col2'])

df_list

col1  col2
0    10    20
1    30    40
2    50    60

**2. Create DataFrame a Dictionary of Lists**
- You can create a DataFrame from a Dictionary where,
    - `key` represent column names.
    - `value` represent list containing data for each column.

In [9]:
# define dictionary of lists
data_dict = {
    'Name': ['Ram', 'Gita', 'Hari'],
    'Age': [10, 25, 20]
}

# create dataframe
df_dict = pd.DataFrame(data_dict)

df_dict

Name  Age
0   Ram   10
1  Gita   25
2  Hari   20

**3. Create DataFrame from a List of Dictionaries**
- You can also create DataFrame from a List of Dictionaries, where,
    - each dictionary corresponds to a row and contains column-value pairs.

In [10]:
# define data list
data_list = [
    {'Name': 'Ram', 'Age': 10},
    {'Name': 'Gita', 'Age': 25},
    {'Name': 'Hari', 'Age': 20}
]

# Create DataFrame
df_list = pd.DataFrame(data_list)

df_list

Name  Age
0   Ram   10
1  Gita   25
2  Hari   20

## Data Ingestion
- Data Ingestion is the process of obtaining and import data for immediate use or storage in a database.
- This repository includes variety of Dataset, so for now we will use that as data source.
- There may be scenario where we need to ingest data from Google Drive as well for that purpose we can use `gdown` library

**5. Create DataFrame from Csv File**
- Use `pd.read_csv()` function

In [11]:
# define path
data_path = '../../Data/HR_comma_sep.csv'

# read data from csv file
df_hr = pd.read_csv(data_path)

df_hr

FileNotFoundError: [Errno 2] No such file or directory: '../../Data/HR_comma_sep.csv'

**5. Create DataFrame from an Excel file**
- Use  `pd.read_excel()` function

In [ ]:
# write your code here

## Viewing Data
- In this sections we will explore methods available to view our loaded DataFrame object.
- **Methods are**
    - head() and tail()
    - shape
    - index and columns
    - dtypes
    - values and to_numpy()
    - info()  
    - describe()

In [ ]:
# use head() to view top 5 rows
df_hr.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [ ]:
# use tail() to view last 5 rows
df_hr.tail()

satisfaction_level  last_evaluation  number_project  \
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department salary  
14994                      0    support    low  
14995                      0    support    low  
14996                      0    support    low  
14997                      0    support    low  
14998                      0    support    low

In [ ]:
# Check dataframe shape
# Hint: df.shape

df_hr.shape

(14999, 10)

In [ ]:
# extract dataframe index
# Hint: df.index

df_hr.index

RangeIndex(start=0, stop=14999, step=1)

In [ ]:
# extract  columns
# Hint: df.columns

df_hr.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [ ]:
# get column types
# Hint: df.dtypes

df_hr.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                       int64
promotion_last_5years      int64
Department                object
salary                    object
dtype: object

- We can see, 
    - 2 columns are of type float64
    - 6 columns are of type int64
    - 2 columns are of type object

In [ ]:
# convert pandas dataframe to numpy array
# Hint: df.values
df_hr.values

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

In [ ]:
# convert pandas dataframe to numpy array
# Hint: df.to_numpy()

df_hr.to_numpy()

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

In [ ]:
# get basic dataframe information
# Hint: df.info()

df_hr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


- Dataframe occupies 1.1+ MB memory
- Columns are of types: float64, int64, object
- Total rows = 14999
- No any columns contains null values

In [ ]:
# Descriptive Statistics
# Hint: df.describe()

df_hr.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

- We get the summary of each dataframe column easily without the need for explicit computation.
- By default it includes only the numerical columns.
- let's get description  of object type only in below cells.

In [ ]:
# describe object type, exclude int, and float
# Hint: df.describe(exclude=['int', 'float'])

df_hr.describe(exclude=['int', 'float'])

Department salary
count       14999  14999
unique         10      3
top         sales    low
freq         4140   7316

- For Object type, we get count, unique, top items with frequency.

## Data Selection
- This section covers different techniques to access columns, slice dataframe, get unique column values.
    - select columns (using [], and dot(.) operator)
    - iloc
    - loc 
    - unique() and nunique()
    - value_counts()

**Select Columns**

In [ ]:
# Select single column

# without using dot operator
print(df_hr['satisfaction_level'])

print('------------------------------------------')

# using dot operator
print(df_hr.satisfaction_level)

0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64
------------------------------------------
0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64


In [ ]:
# Select multiple columns

columns_to_select = ['satisfaction_level', 'left']

df_hr[columns_to_select]

satisfaction_level  left
0                    0.38     1
1                    0.80     1
2                    0.11     1
3                    0.72     1
4                    0.37     1
...                   ...   ...
14994                0.40     1
14995                0.37     1
14996                0.37     1
14997                0.11     1
14998                0.37     1

[14999 rows x 2 columns]

**iloc**  
- ways to slice pandas dataframe via integer value.
- It allows you to select data from a DataFrame using integer indices for both rows and columns.
- 1st row is 0, 2nd row is 1, and so on.
- 1st column is 0, 2nd column is 1, and so on.
- **`syntax: df.iloc[row_index, column_index]`**

In [ ]:
# select single element using iloc
# value at 1st row and 1st column

df_hr.iloc[0, 0]

0.38

In [ ]:
# slice dataframe
# 1st 2 rows and last 3 columns

df_hr.iloc[:3, -3:]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

**loc**
- `loc` is used for label-based indexing. 
- Allows us to slice dataframe using row label and column label instead of integer value as in loc.  
- **`syntax: df.loc[row_label, column_label]`**

In [ ]:
# select single element using loc
# value at 1st row and 1st column

df_hr.loc[0, 'satisfaction_level']

0.38

In [ ]:
# slice dataframe
# 1st 2 rows and last 3 columns

row_label = [0, 1, 2]
column_label = ['promotion_last_5years', 'Department', 'salary']

df_hr.loc[row_label, column_label]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

**unique() and nunique()**
- `unique()` function to get unique item names. You need to have Series data structure not DataFrame.
- `nunique()` function to get number of unique item names. You need to have Series data structure not DataFrame.

In [ ]:
# get unique item names of column "left"
# Hint: df['col_name'].unique()

df_hr['left'].unique()

array([1, 0])

In [ ]:
# get unique item count of column "left"
# Hint: df['col_name'].nunique()

df_hr['left'].nunique()

2

- This means `left` column has 2 unique values,
    - 1: indicates employee leaving company
    - 0: indicates employee staying in company

**value_counts()**
- This method combines both the idea of unique() and nunique() and gives data count in each unique category.


In [ ]:
# value_counts left column
# Hint: df['left].value_counts()

df_hr['left'].value_counts()

0    11428
1     3571
Name: left, dtype: int64

## Aggregations
- `Methods are:`
    - mean()
    - std()
    - sum()
    - max()
    - min()

In [ ]:
# 1. compute mean of dataframe (axis = 0)

df_hr.mean(axis=0)

/tmp/ipykernel_3509229/4223322009.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_hr.mean(axis=0)


satisfaction_level         0.612834
last_evaluation            0.716102
number_project             3.803054
average_montly_hours     201.050337
time_spend_company         3.498233
Work_accident              0.144610
left                       0.238083
promotion_last_5years      0.021268
dtype: float64

In [ ]:
# 2. compute standard deviation

df_hr[['satisfaction_level', 'last_evaluation']].std()

satisfaction_level    0.248631
last_evaluation       0.171169
dtype: float64

In [ ]:
df_hr.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [ ]:
# 3. sum()

# get total work accident
print(df_hr['Work_accident'].sum())

2169


In [ ]:
# get total work accident and left number

df_hr[['Work_accident', 'left']].sum()

Work_accident    2169
left             3571
dtype: int64

In [ ]:
# get maximum employee satisfaction level
df_hr['satisfaction_level'].max()

1.0

In [ ]:
# get minimum employee satisfaction level
df_hr['satisfaction_level'].min()

0.09

- It means minimum employee satisfaction is 0.09 but not 0.
- Maximum employee satisfaction is 1 i.e. some employee are fully satisfied with the company.

## Missing Data
- In this sections we will explore methods available to find missing data and handle missing data in our data.
- `Methods are:`
    - isnull() or isna()
    - notnull() or notna()
    - dropna()
    - fillna()

**isnull() or isna()**
- Returns True if DataFrame values are null
- Dataframe method to get nan values from columns.
- It will considered null values if there are `nan` which is python default marker for null values.
- you can use `df.isnull()` methods which is simply an alias of `df.isna()`

In [ ]:
# get null values column wise.

df_hr.isnull()

satisfaction_level  last_evaluation  number_project  \
0                   False            False           False   
1                   False            False           False   
2                   False            False           False   
3                   False            False           False   
4                   False            False           False   
...                   ...              ...             ...   
14994               False            False           False   
14995               False            False           False   
14996               False            False           False   
14997               False            False           False   
14998               False            False           False   

       average_montly_hours  time_spend_company  Work_accident   left  \
0                     False               False          False  False   
1                     False               False          False  False   
2                     False               False          False  False   
3                     False               False          False  False   
4                     False               False          False  False   
...                     ...                 ...            ...    ...   
14994                 False               False          False  False   
14995                 False               False          False  False   
14996                 False               False          False  False   
14997                 False               False          False  False   
14998                 False               False          False  False   

       promotion_last_5years  Department  salary  
0                      False       False   False  
1                      False       False   False  
2                      False       False   False  
3                      False       False   False  
4                      False       False   False  
...                      ...         ...     ...  
14994                  False       False   False  
14995                  False       False   False  
14996                  False       False   False  
14997                  False       False   False  
14998                  False       False   False  

[14999 rows x 10 columns]

In [ ]:
# sum along axis = 0

df_hr.isnull().sum()

satisfaction_level       0
last_evaluation          0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
Department               0
salary                   0
dtype: int64

This means there are no any values in our DataFrame.

**notnull() or notna()**
- Returns True if DataFrame values are not null.
- You can use method `df.notnull()` which is simply an alias of `df.notna()`

In [ ]:
# get not null 

df_hr.notnull()

satisfaction_level  last_evaluation  number_project  \
0                    True             True            True   
1                    True             True            True   
2                    True             True            True   
3                    True             True            True   
4                    True             True            True   
...                   ...              ...             ...   
14994                True             True            True   
14995                True             True            True   
14996                True             True            True   
14997                True             True            True   
14998                True             True            True   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                      True                True           True  True   
1                      True                True           True  True   
2                      True                True           True  True   
3                      True                True           True  True   
4                      True                True           True  True   
...                     ...                 ...            ...   ...   
14994                  True                True           True  True   
14995                  True                True           True  True   
14996                  True                True           True  True   
14997                  True                True           True  True   
14998                  True                True           True  True   

       promotion_last_5years  Department  salary  
0                       True        True    True  
1                       True        True    True  
2                       True        True    True  
3                       True        True    True  
4                       True        True    True  
...                      ...         ...     ...  
14994                   True        True    True  
14995                   True        True    True  
14996                   True        True    True  
14997                   True        True    True  
14998                   True        True    True  

[14999 rows x 10 columns]

In [ ]:
# sum along axis = 0

df_hr.notnull().sum(axis=0)

satisfaction_level       14999
last_evaluation          14999
number_project           14999
average_montly_hours     14999
time_spend_company       14999
Work_accident            14999
left                     14999
promotion_last_5years    14999
Department               14999
salary                   14999
dtype: int64

- This also shows that no any columns contains nan values

**dropna()**
- Removes rows (by default axis = 0) with missing values.
- You can also removes columns with missing values (setting axis = 1) 

In [ ]:
# remove rows with missing values
# Hint: df.dropna()

df_hr.dropna()

satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department  salary  
0                          0      sales     low  
1                          0      sales  medium  
2                          0      sales  medium  
3                          0      sales     low  
4                          0      sales     low  
...                      ...        ...     ...  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low  

[14999 rows x 10 columns]

**fillna()**
- Fill missing values with given values
- You can use `df.fillna(value)` function

In [ ]:
# fill missing values with 5
# Hint: df.fillna(value)

df_hr.fillna(5)

satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department  salary  
0                          0      sales     low  
1                          0      sales  medium  
2                          0      sales  medium  
3                          0      sales     low  
4                          0      sales     low  
...                      ...        ...     ...  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low  

[14999 rows x 10 columns]

- Since there are no any missing values in our dataset, so dropna() and fillna() is not useful

## Data Manipulation
- `Methods are:`
    - replace()
        - replace values from dataframe
        - df.replace({'old_val': 'new_val', 'old_val_2': 'new_val_2',...})
    - astype()
        - change dataframe column types
        - df.astype({'col1': 'new_type', 'col2': 'new_type2',......})
    - drop()
        - Drop column
        - df.drop(col_name, axis=1)
    - sort_values()
        - sort values either in ascending order or descending order.
        - df.sort_values(by='col', ascending=True, inplace=True) --> SOrt in ascending order
        - df.sort_values(by='col', ascending=False, inplace=True) --> SOrt in descending order
    - groupby()
        - Group DataFrame by one or more columns
        - df.groupby(['col1', 'col2', ..]).agg({'col': 'sum', 'col', 'count'})
    - apply()
        - applies user defined function or lambda function to the data.
        - df.apply(func, axis=0)
            - func: The function to apply to each column (default) or row in the DataFrame
            - axis: The axis along which the function will be applied.

In [ ]:
# Replace values from "left column"
# 1 by yes and 0 by no
# Hint: df.replace({'old_val': 'new_val', 'old_val_2': 'new_val_2'})

df_hr['left'].replace({1: 'yes', 0: 'no'}, inplace=True)
df_hr['left'].value_counts()

no     11428
yes     3571
Name: left, dtype: int64

In [ ]:
# Change column types in dataframe
# convert "left" column to type "bool"

df_typecast = df_hr.astype({'left': 'bool'})
df_typecast.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0  True                      0      sales   
1                   6              0  True                      0      sales   
2                   4              0  True                      0      sales   
3                   5              0  True                      0      sales   
4                   3              0  True                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [ ]:
# Drop column
# example: drop column named salary
# Hint: df.drop('col_name', axis=1)

df_drop = df_hr.drop('salary', axis=1)
df_drop.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident left  promotion_last_5years Department  
0                   3              0  yes                      0      sales  
1                   6              0  yes                      0      sales  
2                   4              0  yes                      0      sales  
3                   5              0  yes                      0      sales  
4                   3              0  yes                      0      sales

In [ ]:
# sort values by Work accident
df_hr.sort_values(by='Work_accident', ascending=False)

satisfaction_level  last_evaluation  number_project  \
8078                 0.77             0.91               5   
3570                 0.50             0.60               3   
1784                 0.43             0.51               2   
6589                 0.64             0.64               3   
11800                0.67             0.91               2   
...                   ...              ...             ...   
5310                 0.63             0.87               5   
5311                 0.86             0.75               5   
5312                 0.80             0.79               5   
5313                 0.55             0.58               5   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident left  \
8078                    268                   3              1   no   
3570                    270                   3              1   no   
1784                    149                   3              1  yes   
6589                    234                   3              1   no   
11800                   245                   2              1   no   
...                     ...                 ...            ...  ...   
5310                    248                   2              0   no   
5311                    157                   4              0   no   
5312                    240                   2              0   no   
5313                    262                   3              0   no   
14998                   158                   3              0  yes   

       promotion_last_5years  Department  salary  
8078                       0          IT  medium  
3570                       0  management  medium  
1784                       0  accounting  medium  
6589                       0       sales    high  
11800                      1       sales  medium  
...                      ...         ...     ...  
5310                       0   technical     low  
5311                       0     support     low  
5312                       0     support     low  
5313                       0     support  medium  
14998                      0     support     low  

[14999 rows x 10 columns]

In [ ]:
# groupby
# Department wise employee promotion
#
# Hint: df.groupby('Department').agg({'promotion_last_5years': 'sum'})

df_depart_left = df_hr.groupby('Department').agg({'promotion_last_5years': 'sum'})
df_depart_left

promotion_last_5years
Department                        
IT                               3
RandD                           27
accounting                      14
hr                              15
management                      69
marketing                       43
product_mng                      0
sales                          100
support                         20
technical                       28

In [ ]:
# apply
# change left column to bool

df_hr['left'] = df_hr['left'].apply(lambda x: bool(x))
df_hr.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                        bool
promotion_last_5years      int64
Department                object
salary                    object
dtype: object

- we can see column "left" is now changed to boolean type.
- Similary we can apply apply function either in series or in a whole dataframe based on our use cases.

## Data Join
- `Types are:`
    1. Inner Join
        - pd.merge(df1, df2, on='col_name', how='inner')
    2. Left Outer Join
        - pd.merge(df1, df2, on='col_name', how='left')
    3. Right Outer Join
        - pd.merge(df1, df2, on='col_name', how='right')
    4. Full Outer Join
        - pd.merge(df1, df2, on='col_name', how='outer')

In [ ]:
df_hr.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [ ]:
# reset index to get employee id

df_hr_ = df_hr.reset_index()

# rename col index to emp_id
df_hr_.columns.values[0] = "emp_id"

df_hr_.head()



emp_id  satisfaction_level  last_evaluation  number_project  \
0       0                0.38             0.53               2   
1       1                0.80             0.86               5   
2       2                0.11             0.88               7   
3       3                0.72             0.87               5   
4       4                0.37             0.52               2   

   average_montly_hours  time_spend_company  Work_accident  left  \
0                   157                   3              0     1   
1                   262                   6              0     1   
2                   272                   4              0     1   
3                   223                   5              0     1   
4                   159                   3              0     1   

   promotion_last_5years Department  salary  
0                      0      sales     low  
1                      0      sales  medium  
2                      0      sales  medium  
3                      0      sales     low  
4                      0      sales     low

In [ ]:
# seggregate emp_id and department
# later we will join it

df_depart = df_hr_[['emp_id', 'Department']]

# drop department from original table
df_hr_.drop('Department', axis=1, inplace=True)

df_hr_.head()

emp_id  satisfaction_level  last_evaluation  number_project  \
0       0                0.38             0.53               2   
1       1                0.80             0.86               5   
2       2                0.11             0.88               7   
3       3                0.72             0.87               5   
4       4                0.37             0.52               2   

   average_montly_hours  time_spend_company  Work_accident  left  \
0                   157                   3              0     1   
1                   262                   6              0     1   
2                   272                   4              0     1   
3                   223                   5              0     1   
4                   159                   3              0     1   

   promotion_last_5years  salary  
0                      0     low  
1                      0  medium  
2                      0  medium  
3                      0     low  
4                      0     low

**1. inner join**

In [ ]:
# inner join based on employee id to get Department information

df_inner = pd.merge(df_hr_, df_depart, on='emp_id', how='inner')

df_inner.head()

emp_id  satisfaction_level  last_evaluation  number_project  \
0       0                0.38             0.53               2   
1       1                0.80             0.86               5   
2       2                0.11             0.88               7   
3       3                0.72             0.87               5   
4       4                0.37             0.52               2   

   average_montly_hours  time_spend_company  Work_accident  left  \
0                   157                   3              0     1   
1                   262                   6              0     1   
2                   272                   4              0     1   
3                   223                   5              0     1   
4                   159                   3              0     1   

   promotion_last_5years  salary Department  
0                      0     low      sales  
1                      0  medium      sales  
2                      0  medium      sales  
3                      0     low      sales  
4                      0     low      sales

## Save Data
- Like Read, Pandas provides several function to save data to various file formats such as CSV, Excel, JSON, and so on.
- **Save CSV**
    - `df.to_csv(csv_file_path, index=False)`
        - csv_file_path: csv path where file needs to be saved: e.g. test.csv
        - index=False: means do not save index data.

- **Save Excel**
    - `df.to_excel(excel_file_path, index=False)`

In [ ]:
# write your program here
# save department data

df_depart.to_csv('../../Data/deparment.csv', index=False)